### DISCUSSION TITLE
Any tips for chezmoi on VS Code?

### DISCUSSION THREAD

**Discussion by felipecrs** [twpayne/chezmoi#1426] [Show and tell]
I have one:

https://marketplace.visualstudio.com/items?itemName=Tim-Koehler.helm-intellisense

Then:

```jsonc
// .vscode/settings.json
{
  "files.associations": {
    ".chezmoi*.yaml": "helm"
  }
}
```

Result:

https://user-images.githubusercontent.com/29582865/133840106-3941d18c-1064-4e04-a3bc-d90a31483bf7.mp4

Refs https://github.com/chezmoi/dotfiles/pull/8

**Comment by felipecrs** [comment]
For scripts, I have been doing things like:

```bash
# {{ if .is_wsl }}
# The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
if service docker status 2>&1 | grep -q "is not running"; then
  wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
fi
# {{ end }}
```

To prevent my ShellCheck and other shell extensions to freak out. The pitfall is that the `#` sign remains in the code after rendering the template. @twpayne do you know if there is anything that can be written to the template to wipe out the leadings `#` from the final script?

**Reply by twpayne** [reply]
> @twpayne do you know if there is anything that can be written to the template to wipe out the leadings `#` from the final script?

I don't think this is possible. Fundamentally, this is a text/template template and not a shell script. text/template has some markers for stripping whitespace, but not anything else as far as I know.

I would suggest instead using `chezmoi execute-template` (or `chezmoi cat`) to get the generated shell script and feed that into shellcheck something like:

```console
$ shellcheck <(chezmoi execute-template $PATH_TO_TEMPLATE_IN_SOURCE_DIR)
```

Insider tip: `chezmoi cat` also works with scripts. You need to strip any chezmoi prefixes and suffixes (e.g. `run_once_` and `.tmpl`), so if you have a script called `run_once_foo.sh.tmpl` then you can also do:

```console
$ shellcheck <(chezmoi cat foo.sh)
```

**Reply by felipecrs** [reply]
That's cool, but I wanted to use ShellCheck in VS Code and not in the command-line. BTW, the following works:

```console
$ cat home/run_once_before_1_setup_apt_packages.sh.tmpl | chezmoi execute-template | code -
```

![image](https://user-images.githubusercontent.com/29582865/135650460-815771d7-fe36-4049-87cb-51cf2af341cb.png)

From the screenshot you may notice that the cat command did not work...

**Reply by felipecrs** [reply]
But I will probably stick with the `#` signs yet. They do not cause any issues... they are just ugly.

**Reply by bradenhilton** [reply]
<details>
<summary>Incorrect solution</summary>

<del>

Would anyone like an ugly hacky workaround for an ugly hacky workaround?

If you're using an action that returns some kind of output on the same line as the comment (`template ...`/`include ..`/`output ...` etc.), you can prepend the action with a carriage return via `{{ print "\r" }}` to overwrite the comment:

`example.ps1.tmpl`:

```
# ...

# {{ print "\r" }}{{ template "helpers/powershell/Read-Stdin.ps1" }}
    
# ...
```

```console
❯ Get-Content "example.ps1.tmpl" | chezmoi execute-template
# ...

function Read-Stdin {
    <#
    .SYNOPSIS
        Reads stdin to a string.
    .OUTPUTS
        System.String.
        A string containing the contents of stdin.
    #>

    $stdin = [System.Console]::In
    return $stdin.ReadToEnd()
}

# ...
```

If you're using an action that **does not** return some kind of output on the same line as the comment (template variable assignment, multi-line `if ...`/`range ...` etc.), you need to remove trailing whitespace on the action lines (not entirely sure why this works):

`example.sh.tmpl`:

```
# ...

# {{ print "\r" }}{{ if .is_wsl -}}
# The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
if service docker status 2>&1 | grep -q "is not running"; then
    wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
fi
# {{ print "\r" }}{{ end -}}

# ...
```

```console
❯ Get-Content "example.sh.tmpl" | chezmoi execute-template
# ...

# The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
if service docker status 2>&1 | grep -q "is not running"; then
    wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
fi
# ...
```

You can fix the whitespace by prepending a newline to the `end` line:

```
# ...

# {{ print "\r" }}{{ if .is_wsl -}}
# The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
if service docker status 2>&1 | grep -q "is not running"; then
    wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
fi

# {{ print "\r" }}{{ end -}}

# ...
```

```console
❯ Get-Content "example.sh.tmpl" | chezmoi execute-template
# ...

# The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
if service docker status 2>&1 | grep -q "is not running"; then
    wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
fi

# ...
```

</del>

</details>

**Reply by felipecrs** [reply]
This is amazing!

**Reply by twpayne** [reply]
Doesn't the carriage return `\r` just return the cursor to the start of the line, causing whatever comes after the `\r` to overwrite whatever was previously displayed on the time when the template results are printed? i.e. I don't think this overwrites the comments in the template result, it just makes them not visible when you look at the template output in a terminal, for example:

```console
$ chezmoi execute-template '{{ printf "abcde\r123" }}'
123%
$ chezmoi execute-template '{{ printf "abcde\r123" }}' | xxd
00000000: 6162 6364 650d 3132 33                   abcde.123
```

Maybe this is what you want...

**Reply by bradenhilton** [reply]
@twpayne It would seem that you are correct. I could have sworn I got it to work on the template result at one point but I went through a lot of iterations in a short amount of time, either that or it was just general Windows strangeness. How frustrating!

**Reply by bradenhilton** [reply]
An imperfect solution would be to add support for directive comments which are evaluated before the template is parsed. Maybe `chezmoi:delims *left* *right*` could be used to set https://pkg.go.dev/text/template#Template.Delims, so one could use `chezmoi:delims "#{{" "}}"` or `chezmoi:delims "//{{" "}}"`. Or the typical replacers `replace`/`regexReplaceAll` could be used e.g. `chezmoi:replace "# {{" "{{"` . Seems like it could be a lot of work just to remove some empty comments in the template result though, assuming it's even possible. The directives would have to be at the top of the file and be consumed so they don't appear in the result to preserve things like shebangs.

It just occurred to me that mandating directives be placed at the top of the file could result in ShellCheck complaining that the shebang is not the first line of the file. Perhaps the directives should be placed at the bottom of the file instead?

**Reply by twpayne** [reply]
As an alternative, exposing template data as environment variables (work in progress in #2418) would avoid the need to use templates for your scripts, assuming your scripting language is capable of reading environment variables. I'm thinking something along the lines of:

Config file:
```toml
[env]
    IS_WSL = "true" # set $IS_WSL=true in scripts
```

Script (does not need to be a template, now syntactically valid and so can be checked with ShellCheck):

```bash
if "${IS_WSL}" == "true"; then
  # The boot.command in /etc/wsl.conf is only for Windows 11, so this workaround is needed
  if service docker status 2>&1 | grep -q "is not running"; then
    wsl.exe -d "${WSL_DISTRO_NAME}" -u root -e /usr/sbin/service docker start >/dev/null 2>&1
  fi
fi
```

Note that templates are still needed for config files that are not scripts.

**Reply by PaulRBerg** [reply]
It looks like these environment variables have been implemented. But I cannot find the documentation for them. The docs only mention this:

<img width="400" alt="CleanShot 2025-10-08 at 18 35 44@2x" src="https://github.com/user-attachments/assets/d5b21ec2-103d-4970-a3c2-4eb9c975421f" />

Where can I see the full list, @twpayne?

I'd like to know if `CHEZMOI_HOSTNAME` is set, too.

**Reply by felipecrs** [reply]
I found these references in documentation:

https://www.chezmoi.io/reference/configuration-file/hooks/
https://www.chezmoi.io/user-guide/use-scripts-to-perform-actions/#set-environment-variables

A documented list seems not to exist yet indeed.

**Reply by felipecrs** [reply]
But maybe this helps:

```console
❯ chezmoi cd

❯ printenv | grep CHEZMOI
CHEZMOI=1
CHEZMOI_ARCH=amd64
CHEZMOI_CACHE_DIR=/home/felipecrs/.cache/chezmoi
CHEZMOI_COMMAND_DIR=/home/felipecrs/repos/dotfiles
CHEZMOI_GID=1000
CHEZMOI_GROUP=felipecrs
CHEZMOI_SOURCE_DIR=/home/felipecrs/.dotfiles/home
CHEZMOI_COMMAND=cd
CHEZMOI_EXECUTABLE=/home/linuxbrew/.linuxbrew/Cellar/chezmoi/2.65.2/bin/chezmoi
CHEZMOI_FQDN_HOSTNAME=FELIPE-DESKTOP
CHEZMOI_OS=linux
CHEZMOI_CONFIG_FILE=/home/felipecrs/.config/chezmoi/chezmoi.yaml
CHEZMOI_DEST_DIR=/home/felipecrs
CHEZMOI_WORKING_TREE=/home/felipecrs/.dotfiles
CHEZMOI_ARGS=chezmoi cd
CHEZMOI_HOME_DIR=/home/felipecrs
CHEZMOI_HOSTNAME=FELIPE-DESKTOP
CHEZMOI_UID=1000
CHEZMOI_USERNAME=felipecrs
CHEZMOI_VERBOSE=1
CHEZMOI_VERSION_BUILT_BY=Homebrew
CHEZMOI_VERSION_COMMIT=57705baac6b3977053d67963f75aa68540b5636a
CHEZMOI_VERSION_DATE=2025-09-23T20:56:02Z
CHEZMOI_VERSION_VERSION=2.65.2
CHEZMOI_KERNEL_VERSION=#1 SMP PREEMPT_DYNAMIC Thu Jun  5 18:30:46 UTC 2025
CHEZMOI_KERNEL_OSRELEASE=6.6.87.2-microsoft-standard-WSL2
CHEZMOI_KERNEL_OSTYPE=Linux
CHEZMOI_OS_RELEASE_ID_LIKE=debian
CHEZMOI_OS_RELEASE_SUPPORT_URL=https://help.ubuntu.com/
CHEZMOI_OS_RELEASE_BUG_REPORT_URL=https://bugs.launchpad.net/ubuntu/
CHEZMOI_OS_RELEASE_LOGO=ubuntu-logo
CHEZMOI_OS_RELEASE_NAME=Ubuntu
CHEZMOI_OS_RELEASE_ID=ubuntu
CHEZMOI_OS_RELEASE_UBUNTU_CODENAME=noble
CHEZMOI_OS_RELEASE_HOME_URL=https://www.ubuntu.com/
CHEZMOI_OS_RELEASE_PRIVACY_POLICY_URL=https://www.ubuntu.com/legal/terms-and-policies/privacy-policy
CHEZMOI_OS_RELEASE_VERSION_ID=24.04
CHEZMOI_OS_RELEASE_VERSION=24.04.3 LTS (Noble Numbat)
CHEZMOI_OS_RELEASE_VERSION_CODENAME=noble
CHEZMOI_OS_RELEASE_PRETTY_NAME=Ubuntu 24.04.3 LTS
CHEZMOI_SUBSHELL=1
```

**Comment by shadycuz** [comment]
@felipecrs bro, about twice a week, I stumble upon a git issue where you created or commented in it. We like use the same tech stack for sure.
